Setup NodeJS Structure

simplytunde opened the pull request:
The PR adds the NodeJS backend structure, docker setup.
closes https://github.com/KanaryStack/aliascheck/issues/25

simplytunde commented on README.md:
````diff
@@ -1,152 +1,34 @@
-
-### Run the Application
-
-
-```sh
-$ python manage.py run
-```
+- [Next.js Documentation](https://nextjs.org/docs) - learn about Next.js features and API.
+- [Learn Next.js](https://nextjs.org/learn) - an interactive Next.js tutorial.
 
-Access the application at the address [http://localhost:5000/](http://localhost:5000/)
+You can check out [the Next.js GitHub repository](https://github.com/vercel/next.js/) - your feedback and contributions are welcome!
 
-### Testing
+## Deploy on Vercel
 
-Without coverage:
+The easiest way to deploy your Next.js app is to use the [Vercel Platform](https://vercel.com/new?utm_medium=default-template&filter=next.js&utm_source=create-next-app&utm_campaign=create-next-app-readme) from the creators of Next.js.
 
-```sh
-$ python manage.py test
-```
-
-With coverage:
-
-```sh
-$ python manage.py cov
-```
-
-Run flake8 on the app:
-
-```sh
-$ python manage.py flake
-```
-
-or
-
-```sh
-$ flake8 project
-```
````
I did not make any change to the README. The README will be updated as part of issue https://github.com/KanaryStack/aliascheck/issues/28

ChubaOraka replied:
Here's [the change your PR makes to the README](https://github.com/KanaryStack/aliascheck/pull/26/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L1-L152). This change occurs because many of the changes you make were done in the root directory and I think the change to the README was unintended.

simplytunde commented on README.md:
```diff
@@ -1,152 +1,34 @@
-# **ALIAS CHECK**
+This is a [Next.js](https://nextjs.org/) project bootstrapped with [`create-next-app`](https://github.com/vercel/next.js/tree/canary/packages/create-next-app).
```
check https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952111893

ChubaOraka replied:
The suggested change to the README blows away [all the change that was made to the README that was intended to onboard new participants](https://github.com/KanaryStack/aliascheck/commit/84e7c75267a7bbbef20433996e505bdf89f36b03) and [replaces it with a boilerplate README](https://github.com/KanaryStack/aliascheck/blob/fcaa7d30086aece64e201e14769a824fbf2bf5e1/README.md).

simplytunde replied:
Good catch. Unintended. I will revert so we can probably update the README as part this issue  https://github.com/KanaryStack/aliascheck/issues/28

ChubaOraka commented on .dockerignore:
```diff
@@ -0,0 +1,8 @@
+.dockerignore
+*.md
+node_modules
+build
+client
+.github
+dist
+.env
```
I believe this file should be in the [`server` folder](https://github.com/KanaryStack/aliascheck/tree/68233d431d50356bb8c4d83134b99f99f0f6536f/server).

simplytunde replied:
The `.dockerignore` location is guided by the location of the dockerfiles. See https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952086936

ChubaOraka replied:
Please refer to [my earlier comment addressing the preferred location of the `Dockerfile`](https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952487845).

simplytunde commented on tsconfig.build.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "extends": "./tsconfig",
+    "exclude": ["**/*.test.*", "**/__mocks__/*", "**/__tests__/*"]
+}
```
I intend for the configuration to be in the root folder. For example, checkout https://github.com/GeekyAnts/express-typescript

simplytunde replied:
@ChubaOraka  Also take a look at some of the largest nodejs projects https://github.com/strapi/strapi https://github.com/TryGhost/Ghost https://github.com/laurent22/joplin. Shared configuration should be in the root folder. Another example is this large project built with nextjs https://github.com/wulkano/Kap. Common configs like tsconfig.json stays in the root folder.

ChubaOraka replied:
Those projects seem to be backend-first projects where the backend framework effectively clobbers the frontend and the frontend files are all in one subdirectory of the backend project. In some of these backend frameworks, there is an inbuilt function to compile the frontend libraries (e.g. React) like is done in [this Rails project](https://github.com/do-community/react_rails_recipe) where most of the frontend React files [are in the `javascript` subfolder](https://github.com/do-community/react_rails_recipe/tree/master/app/javascript). That's one way to go, but I think [our project](https://github.com/KanaryStack/aliascheck/) can be more amenable to separating backend and frontend, whether we are using Flask or Node at the backend. Especially with the case of Node, I think our folder structure can look more better separated like [this project](https://github.com/ChubaOraka/minthack).

To be sure, common files can be placed at the root directory; I just think that placing them in a folder named `shared-configs` makes it more explicit and easily apparent in the code when the files in the [`server`](https://github.com/KanaryStack/aliascheck/tree/68233d431d50356bb8c4d83134b99f99f0f6536f/server) and [`client`](https://github.com/KanaryStack/aliascheck/tree/68233d431d50356bb8c4d83134b99f99f0f6536f/client) folders refer to them.

simplytunde replied:
ok I will place the common files like `gitignore`, ts configs etc at the root directory and move some client and server specific files into each respectively.

simplytunde replied:
I will really like us to follow the repo style of these popular projects.

ChubaOraka replied:
I understand. But you might hurt the frontenders' feelings.

simplytunde commented on tsconfig.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "compilerOptions": {
+    "target": "es2015",
+    "lib": ["dom", "dom.iterable", "esnext"],
+    "types": ["node", "jest"],
+    "allowJs": false,
+    "skipLibCheck": true,
+    "strict": true,
+    "forceConsistentCasingInFileNames": true,
+    "noEmit": false,
+    "esModuleInterop": true,
+    "module": "commonjs",
+    "moduleResolution": "node",
+    "resolveJsonModule": true,
+    "isolatedModules": true,
+    "jsx": "preserve",
+    "incremental": false,
+    "outDir": "./dist"
+  },
+  "compileOnSave": true,
+  "include": ["server.ts", "server/**/*.ts", "server/**/*.tsx"],
+  "exclude": ["node_modules"]
+}
```
https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952112875

ChubaOraka replied:
Please refer to [my earlier comment addressing separation of concerns](https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952464367).

simplytunde commented on .eslintrc:
```diff
@@ -0,0 +1,15 @@
+{
+    "root": true,
+    "parser": "@typescript-eslint/parser",
+    "plugins": [
+        "@typescript-eslint"
+    ],
+    "extends": [
+        "eslint:recommended",
+        "plugin:@typescript-eslint/eslint-recommended",
+        "plugin:@typescript-eslint/recommended"
+    ],
+    "rules": {  //For list of rules, https://eslint.org/docs/latest/rules/
+        "no-console": 2 // Remember, this means error!  
+    }
+}
```
checkout https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952112875

simplytunde commented on .eslintignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+dist
```
checkout https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952112875

simplytunde commented on server.ts:
```diff
@@ -0,0 +1,10 @@
+import {logger, app} from './server/app';
+import * as dotenv from 'dotenv';
+
+dotenv.config();
+const port: string | undefined = process.env.PORT;
+app.listen(port, () => {
+  logger.info(`⚡️[server]: Server is running at https://localhost:${port}`);
+});
+
+module.exports = app;
```
The `server.ts` can be moved into the server folder. I will need to make rearrangment.

simplytunde commented on package-lock.json:
I can move this into the `server` folder since package.json is unique to both

simplytunde commented on .gitignore:
```diff
@@ -0,0 +1,38 @@
+# typescript
+*.tsbuildinfo
+.env
+server.js
```
Typescript is the supported lang hence we will most likely not going back to JS. If for someone reason we go back to JS, it quite easy to add to gitignore

simplytunde commented on .gitignore:
```diff
@@ -0,0 +1,38 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
+
+# dependencies
+/node_modules
+/.pnp
+.pnp.js
+
+# testing
+/coverage
+
+# next.js
+/.next/
+/out/
+
+# production
+/build
+
+# misc
+.DS_Store
+*.pem
+
+# debug
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+.pnpm-debug.log*
+
+# local env files
+.env*.local
+
+# vercel
+.vercel
+
+# typescript
+*.tsbuildinfo
+.env
+server.js
+dist/
```
`.gitgnore` should always be in the root folder since it can be applied even to deeply nested subfolders.

simplytunde commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM node:16.17.0-alpine AS builder
+WORKDIR /usr
+# COPY package.json ./
+# COPY tsconfig.json ./
+COPY . .
+RUN ls -a
+RUN npm install
+RUN npm run build
+
+## this is stage two , where the app actually runs
+
+FROM node:16.17.0-alpine
+WORKDIR /usr
+COPY package.json .
+RUN npm install --only=production
+COPY --from=builder /usr/dist .
+RUN npm install pm2 -g
+EXPOSE ${PORT}
+CMD ["pm2-runtime","server.js"]
```
Agreed. I can make a simpler one for `dev`

simplytunde commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM node:16.17.0-alpine AS builder
+WORKDIR /usr
+# COPY package.json ./
+# COPY tsconfig.json ./
+COPY . .
+RUN ls -a
+RUN npm install
+RUN npm run build
+
+## this is stage two , where the app actually runs
+
+FROM node:16.17.0-alpine
+WORKDIR /usr
+COPY package.json .
+RUN npm install --only=production
+COPY --from=builder /usr/dist .
+RUN npm install pm2 -g
+EXPOSE ${PORT}
+CMD ["pm2-runtime","server.js"]
```
Just like `.gitignore`, this should be in the root folder or another subfolder `docker/`

ChubaOraka replied:
This Dockerfile is intended to setup the Node backend, so it's better it replaces the [Flask Dockerfile](https://github.com/KanaryStack/aliascheck/blob/a0642aeba0f9403bc26cd578178ee910e0269199/backend/Dockerfile) that was at the [`backend` folder](https://github.com/KanaryStack/aliascheck/tree/3f20eaa3aed83626d17d846c0f955a16e3f24737/backend) which you have now renamed [`server`](https://github.com/KanaryStack/aliascheck/tree/23452c0dfc10244ffe0a75907c21ed5e0ff47804/server).

**Side note:**
`.gitignore` is ideally placed at the root directory but [can also be placed at subdirectories](https://labs.consol.de/development/git/2017/02/22/gitignore.html#:~:text=gitignore%20in%20every%20subdirectory.,specific%20to%20your%20local%20copy.).

simplytunde commented on package.json:
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "aliascheck",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "dev": "tsc --project './tsconfig.json' --watch & nodemon dist/server.js",
+    "build": "tsc --project './tsconfig.build.json'",
+    "start": "nodemon dist/server.js",
+    "lint": "eslint . --ext .ts",
+    "test": "jest"
+  },
+  "dependencies": {
+    "dotenv": "^16.0.1",
+    "express": "^4.18.1",
+    "http-status-codes": "^2.2.0",
+    "winston": "^3.8.1"
+  },
+  "devDependencies": {
+    "@types/express": "^4.17.13",
+    "@types/jest": "^28.1.7",
+    "@types/node": "^18.0.0",
+    "@types/react": "18.0.14",
+    "@types/react-dom": "18.0.5",
+    "@types/supertest": "^2.0.12",
+    "@typescript-eslint/eslint-plugin": "^5.33.1",
+    "@typescript-eslint/parser": "^5.33.1",
+    "autoprefixer": "^10.4.7",
+    "eslint": "^8.22.0",
+    "eslint-config-next": "12.1.6",
+    "jest": "^28.1.3",
+    "nodemon": "^2.0.19",
+    "supertest": "^6.2.4",
+    "ts-jest": "^28.0.8",
+    "ts-node": "^10.9.1",
+    "typescript": "^4.7.4"
+  }
+}
```
This will be moved into `server/` since it is not shared

simplytunde commented on package-lock.json:
Will be moved to same location as `package.json`. See https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952093135

simplytunde commented on jest.config.ts:
```diff
@@ -0,0 +1,9 @@
+import type { Config } from "@jest/types"
+
+const config: Config.InitialOptions = {
+  preset: "ts-jest",
+  testEnvironment: "node",
+  verbose: true,
+  automock: true,
+}
+export default config
```
`jest.config.ts` is a shared configuration hence can be in the root folder.

ChubaOraka replied:
Please refer to [my earlier comment addressing separation of concerns](https://github.com/KanaryStack/aliascheck/pull/26#discussion_r952464367).

KiptoonKipkurui commented:
Is there a missing .env file? I cant seem to run the server

simplytunde commented:
> Is there a missing .env file? I cant seem to run the server

Yes,  I have updated the repo to include `.env.sample` and updated the readme with instruction
